Add FMC support to STM32F429i_disc1 board

mkschreder opened the pull request:
Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder commented:
Is there a script that will automatically fix coding style?

gmarull commented on drivers/memc/Kconfig.stm32:
```diff
@@ -3,7 +3,7 @@
 config MEMC_STM32
 	bool "Enable STM32 Flexible Memory Controller (FMC)"
-	depends on SOC_SERIES_STM32H7X
+	depends on SOC_SERIES_STM32H7X || SOC_SERIES_STM32F4X
```
This should be added in a separate commit

mkschreder replied:
Ok. Shall I move it to a new branch and submit a new pull request or force push it?

gmarull replied:
All changes should be made on the same PR so that we can track review progress

mkschreder replied:
I think it's all coming in now in the same PR. I know there was an old PR but you have closed it right?

gmarull commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -103,6 +103,9 @@ static inline int stm32_clock_control_on(const struct device *dev,
 		break;
+	case STM32_CLOCK_BUS_AHB3:
+        LL_AHB3_GRP1_EnableClock(pclken->enr);
+        break; 
```
This should be added in a separate commit

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -44,6 +50,83 @@
 		sw0 = &user_button;
 	};
+
+	fmc: memory-controller@A0000000 {
```
```suggestion
	fmc: memory-controller@a0000000 {
```

addresses are usually written in lower-case

mkschreder replied:
Got it!

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -44,6 +50,83 @@
+		reg = <0xA0000000 0x1000>;
+		clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000001>;
+		label = "STM32_FMC";
+		status = "disabled";
```
All of the above needs to go at soc level

mkschreder replied:
Ok. So all the low level generic stuff on soc level and then timing info and pinctrl on board level since it would depend on ram chip. Correct? Or is it better to leave defaults in place on soc level?

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -44,6 +50,83 @@
+			&fmc_d12_pe15
+			&fmc_d13_pd8
+			&fmc_d14_pd9
+			&fmc_d15_pd10
```
nit: on H747 board pins are written in a more compact form, i.e.

```
<&fmc_sdcke1_pb5 &fmc_sdne1_pb6 &fmc_sdclk_pg8
  &fmc_sdnwe_pc0 ...>;
```

mkschreder replied:
Fixed.

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -44,6 +50,83 @@
+			#address-cells = <1>;
+			#size-cells = <0>;
+			label = "STM32_FMC_SDRAM";
+			status = "disabled";
```
this needs to go at soc level

mkschreder replied:
Fixed.

gmarull reviewed (changes requested):
Please check https://docs.zephyrproject.org/latest/contribute/index.html for contributing guidelines (e.g. commit format, coding conventions, tools, etc.)

mkschreder committed:
drivers: memc: allow using stm32h7 memory controller on stm32f4

The memory controller driver uses HAL library so no changes are needed.
The same driver can be used on STM32F4.

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder committed:
drivers: memc: allow "disabled" nodes

Previous code does not take into account that nodes can be set to
"disabled". It is necessary to skip nodes if they have not been
configured by the board dts which will set them to "okay".

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

gmarull commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -24,4 +25,44 @@
 		};
 	};
+
+	sdram2: sdram@d0000000 {
```
This needs to be added at board level, RAM size can't be specified at SoC level as it's an arbitrary value

mkschreder replied:
Makes sense.

gmarull commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -24,4 +25,44 @@
+					STM32_FMC_SDRAM_MWID_16 STM32_FMC_SDRAM_NB_4
+					STM32_FMC_SDRAM_CAS_2 STM32_FMC_SDRAM_SDCLK_PERIOD_2
+					STM32_FMC_SDRAM_RBURST_ENABLE STM32_FMC_SDRAM_RPIPE_0>;
+				st,sdram-timing = <2 7 4 7 2 2 2>;
```
SoC should not contain such defaults, they can take any value depending on the connected memory/es.

mkschreder replied:
Ok that makes sense. Should it be in stm32f429i_disc1.dts? (moved it there in last commit)

gmarull commented on drivers/memc/memc_stm32_sdram.c:
```diff
@@ -129,5 +129,7 @@ static const struct memc_stm32_sdram_config config = {
 	.banks_len = ARRAY_SIZE(bank_config),
 };
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, st_stm32_fmc_sdram), okay)
```
This is not required

mkschreder replied:
Why is it that if I set nodes to "disabled" then ram is still initialized? It doesn't seem to work unless I add this? But maybe there is some better way. Can you suggest a better solution?

gmarull replied:
Kconfig for SDRAM is selected based on DT status, see https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/memc/Kconfig.stm32#L15. If Kconfig setting is not set, this file will not be compiled https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/memc/CMakeLists.txt#L3

mkschreder replied:
What about if I enable it in my _defconfig? I think it should work regardless of how the setting is set. I actually had to explicitly trace the code in the debugger to figure out why my memory controller was still enabled even though I disabled it in the device tree. So actually it should be possible to disable the node in the device tree only and then it should not be configured in code. For example, if you want to temporarily disable the memory controller and use the pins with some other configuration. Then you would have to disable it in menuconfig but it would be easier to just set the node to status "disabled". Am I missing something here?

gmarull replied:
If you don't need memory controller at all just make sure `CONFIG_MEMC=n` (default) and nothing will be initialized.

mkschreder replied:
Is it not the case that all drivers should honor "status" field? Ie if status = "disabled" then it should not configure a bunch of things that are expected to be disabled? @MaureenHelm can you give us some input on this?

mkschreder replied:
Honestly, for me, it took me by surprise when I set it to disabled and memory still continued to work. It seemed wrong so that's why I added this commit..

mkschreder replied:
@gmarull I notice you are using #if DT_NODE_HAS_STATUS(DT_NODELABEL(sdram2), okay) in the linker script. I think we should be using the same thing in the driver C code but for the memc as well. Otherwise we can end up with disabled nodes that are still initialized which would be kindof weird.

gmarull replied:
In this case it is clearly not necessary as Kconfig entry takes into account DT status, so the file will not get compiled if not needed.

mkschreder replied:
It is still a problem if you enable it in kconfig but your devicetree disables it. If it is disabled in device tree it should rely on compiler to remove dead code by not instantiating it. So if this is not done in otherdrivers then it is an error.

gmarull commented on drivers/memc/memc_stm32.c:
```diff
@@ -59,5 +59,7 @@ static const struct memc_stm32_config config = {
 	.pinctrl_len = ARRAY_SIZE(pinctrl),
 };
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, st_stm32_fmc), okay)
```
This is not required

mkschreder replied:
What tests should I run on the board and how do I run all of them?

gmarull replied:
As of today this check is usually not done. Only multi-instance drivers take into account "okay" devices to initialize the enabled ones.

See for example https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/counter/counter_ll_stm32_rtc.c (single instance).

I'm not sure if there is a defined policy @erwango ?

gmarull reviewed (changes requested):
I'd suggest to look at the FMC PR to see how DT is expected to be used. Also, make sure to run the available test on your board.

gmarull reviewed (changes requested):
There are still a few things to be addressed. I'd suggest to look at the FMC PR https://github.com/zephyrproject-rtos/zephyr/pull/29686 to see how DT is expected to be used, the test that is available, etc.

Note that you should also squash fix commits. This PR should ideally have 4 commits:

1. Add AHB3 clock support on M4 (not related to SDRAM, but required)
2. Enable FMC driver on M4
3. Add FMC definitions at M4 SoC files
4. Enable SDRAM on a particular board

mkschreder commented:
> There are still a few things to be addressed. I'd suggest to look at the FMC PR #29686 to see how DT is expected to be used, the test that is available, etc.
> 
> Note that you should also squash fix commits. This PR should ideally have 4 commits:
> 
> 1. Add AHB3 clock support on M4 (not related to SDRAM, but required)
> 2. Enable FMC driver on M4
> 3. Add FMC definitions at M4 SoC files
> 4. Enable SDRAM on a particular board

I have squashed the commits in latest and also run the test:
scripts/sanitycheck --device-testing --device-serial /dev/ttyUSB0 -p stm32f429i_disc1 -T tests/drivers/memc/stm32_sdram/

Seems to be ok! (also my own test works on the board)

gmarull commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -103,6 +103,9 @@ static inline int stm32_clock_control_on(const struct device *dev,
 		break;
+	case STM32_CLOCK_BUS_AHB3:
+		LL_AHB3_GRP1_EnableClock(pclken->enr);
+		break;
```
commit 631cac912ad141f6442b34885474fa0356b21d62 should not reference memc, e.g. `drivers: clock_control: add ...`

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1.dts:
```diff
@@ -46,6 +52,36 @@
+		&fmc_d15_pd10>;
+};
+
+&fmc_sdram {
```
this should be embedded in `fmc` node (no need to use the label), see the FMC PR.

mkschreder replied:
Yeah I wen't over your pull request. Makes sense.

gmarull commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -24,4 +25,26 @@
+			power-up-delay = <100>;
+			num-auto-refresh = <8>;
+			mode-register = <0x220>;
+			refresh-rate = <1386>;
```
These values are board specific

mkschreder replied:
Got it.

gmarull commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -24,4 +25,26 @@
+		label = "STM32_FMC";
+		status = "disabled";
+
+		fmc_sdram: fmc_sdram {
```
these changes should go into a separate commit.

same node names as H7 should be used to be consistent.

mkschreder replied:
Fixed.

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1_defconfig:
```diff
@@ -23,6 +23,12 @@ CONFIG_PINMUX=y
+CONFIG_MEMC=y
+
+# enable SDRAM
+CONFIG_MEMC_STM32_SDRAM=y
```
There's no need to enable SDRAM by default. Test takes care of that, if there was a sample it should be the sample.

mkschreder replied:
So what about enabling it by default for the board? SDRAM is after all board specific and it's always there. Pins connected to SDRAM can't be used for anything else anyway. Is there a logic behind enabling SDRAM at application level and not board level?

mkschreder replied:
@gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".

gmarull replied:
Adding this setting to defconfig means that even the hello world sample will compile SDRAM when it doesn't need it. Upstream boards may be used on a variety of samples/tests. Each sample/test needs to take care of enabling the required stuff using overlay files.

mkschreder replied:
I think it makes sense to enable support for everything that the board supports at board level. If samples fail it will serveas indication that something is wrong in configuration or that there is a serious conflict. Under normal circumstances, enabling all on board peripherals should not break any samples. (if samples break, it is important to catch these instances asap and fix the conflicts)

mkschreder replied:
Ah! I have a solution @gmarull @erwango! The issue is that things get instantiated even when nodes are not present in device tree. The solution is to make sure things *only* getinstantiated and initialized if the corresponding device tree node is there and is status okay. *All* on board peripherals should be enabled on board level in defconfig but device tree nodes should be disabled and then samples and application just sets devicetree nodes to status okay if they are needed. This ensures working configuration for everything supported by the board AND small code size because linker will eliminate dead code automatically for anything that has not been explicitly instantiated even if it has been compiled. Perfect! @MaureenHelm please comment as well.

gmarull replied:
The problem is that Zephyr boards may be used to run a wide variety of samples, and not each sample has the same requirements. In a real application board where you'll likely run a single firmware it makes sense to enable everything at board level.

I think changing current approach would probably deserve an RFC to discuss the implications/side effects, so I think for now it is better to follow the existing approach to have homogeneity in the tree. 

In Zephyr there is a tight relation between DT-Kconfig because of its static approach. So in most cases it does not make sense to have a Kconfig enabled w/o its DT counterpart, and in fact, you'll likely get some compiler warnings/errors. Regarding DT, I think another approach could be to have all board specific stuff defined in the board DT (e.g. SDRAM timings), but still keep the node disabled and enable as needed on samples/tests overlays. Again, this probably needs further discussion as there may be unforeseen implications.

erwango replied:
> @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".

This should work. Let's take a basic example:
```
$ west build -b nucleo_f429zi samples/hello_world/
-- west build: generating a build system
[...]
-- west build: building application
[1/150] Preparing syscall dependency handling
[145/150] Linking C executable zephyr/zephyr_prebuilt.elf
[150/150] Linking C executable zephyr/zephyr.elf
```
If I try to rebuild without modification on the board, nothing happens
```
$ west build -b nucleo_f429zi samples/hello_world/
ninja: no work to do.
```
Then I update board .dts to disable one uart node:
```
&usart3 {
	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
	current-speed = <115200>;
	/* status = "okay"; */
};
```
Launch the build:
```
$ west build -b nucleo_f429zi samples/hello_world/
[0/1] Re-running CMake...
[106/110] Linking C executable zephyr/zephyr_prebuilt.elf
[110/110] Linking C executable zephyr/zephyr.elf
```

So device tree modifications are detected and we should not end up with wrong binary configuration.
If this happens, then there might be issues at some level.


> Ah! I have a solution @gmarull @erwango! The issue is that things get instantiated even when nodes are not present in device tree.

This should not be the case. Instantiation should happen only based on device tree status. Cf upper.

> The solution is to make sure things only getinstantiated and initialized if the corresponding device tree node is there and is status okay. All on board peripherals should be enabled on board level in defconfig but device tree nodes should be disabled and then samples and application just sets devicetree nodes to status okay if they are needed. This ensures working configuration for everything supported by the board AND small code size because linker will eliminate dead code automatically for anything that has not been explicitly instantiated even if it has been compiled. Perfect! @MaureenHelm please comment as well.

First let me explain the strategy we're applying on this board/application configuration split.
Our preoccupation for default configuration for upstream boards is to ensure upstream applications (samples/tests) portability. This is required for:
- demonstrate zephyr modularity and small footprint
- ease CI testing by ensuring maximal portability of these applications. Running one of these applications on a new board should require minimum modification.

To achieve this, responsibilities are split in 2:
- board configuration is done on board side. It is up to the board to provide the default configuration that should be applied (which i2c instance should be used by default). This recommendation is documented here https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations
- Enabling required subsystems to achieve the required functions (CONFIG_I2C=y, CONFIG_LOG=y, ...) is application responsibility

This way, applications maintainers only have to care about the subsystems they need for their application, also running an application on a new board doesn't require any modification on the application.
On their side, boards maintainers should only care about the peripherals the board supports. No dependencies on application.


Current way of working is applied for the +200 boards supported upstream by zephyr and XX samples/tests. This is not perfect and there are still improvements to be done. 
Also, this strategy has limits and you'll happen to find some boards/foo.overlay within samples/ and tests/. This could be required when it's not possible to provide a default board configuration that satisfies all applications. We're trying to minimize these, but I don't think we'll be able to totally getting rid of these. 


Now, let me come back on your proposal. I'm affraid this would not work in our set up.
Let's say you want to be able to run i2c and spi sample on board foo. Default configuration is done on board side (CONFIG_I2C=y, CONFIG_SPI=y) so the board can support the 2 applications.
In order to run `sample/i2c` you'd need to enable _i2c_y node in `sample/i2c/boards/foo.overlay`.
In order to run `sample/spi` you'd need to enable _spi_x node in `sample/spi/boards/foo.overlay`.
At the end you'll get one board overlay per sample (+200 for now, and keeps increasing) which is huge and hardly maintainable while on boards/ we'll get redundant list of CONFIG_FOO activation, which has quite little value.

gmarull replied:
Thanks for this detailed explanation @erwango!

I think that the case it is confusing is for example the following:

Let's take the `samples/drivers/counter/alarm`. In case of STM32, it will use the `#define TIMER DT_LABEL(DT_INST(0, st_stm32_rtc))` device. If one takes for example `stm32h747i_disco_m7` board, sample builds and runs fine regardless of the state of the `rtc` node in DT. In fact, that board doesn't even reference `rtc`, so it should be disabled. This could be prevented if the driver `DEVICE_AND_API_INIT` is guarded with a `status = "okay"` check as Martin suggests, correct me if I'm wrong.

mkschreder replied:
> > @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".
> 
> This should work. Let's take a basic example:
> 

It works there because uart is instantiated by checking ok status "DT_INST_FOREACH_STATUS_OKAY(STM32_UART_INIT)". The memc code just calls a simpler version DEVICE_DEFINE which I think calls init regardless of device tree status. Which is wrong.

gmarull replied:
> > > @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".
> > 
> > 
> > This should work. Let's take a basic example:
> 
> It works there because uart is instantiated by checking ok status "DT_INST_FOREACH_STATUS_OKAY(STM32_UART_INIT)". The memc code just calls a simpler version DEVICE_DEFINE which I think calls init regardless of device tree status. Which is wrong.

Multi-instance drivers are guarded today because it turns out they all use the `DT_INST_FOREACH_STATUS_OKAY` macro. I see that most, if not all, single-instance drivers like FMC, counter, flash... don't have an equivalent guard. You're right, I think they should also be guarded after all. @erwango are you aware of any reason why it is not done?

mkschreder replied:
@erwango First of all thank you for the really long and detailed explanation. However I need some more clarity here. 

First of all I'm a bit confused with your reply. You say "I'm afraid this would not work in our set up" - which I assumed referred to my reply you quoted - and then you propose that settings like CONFIG_I2C=y, CONFIG_SPI=y should be on board level and defaults should be enabled at board level as well - which is one step further ahead of my point even. I agree with you on that but seem to not be on the same page in this thread. Do we enable I2C and SPI on board level and provide device tree nodes that are enabled by default? Do we provide a generic overlay for sample that enables only aliased devices? Do we even put CONFIG*_STM32* in samples because that would require sample to know which board it is running on? Is that wrong? So which approach is the right one? Is it to enable all board peripherals at board level instead of in sample overlays? 

Secondly, "This should not be the case. Instantiation should happen only based on device tree status. Cf upper.". This is the case unless one uses DT macros that explicitly check for OKAY status. I think this should be made mandatory so that all drivers should use the same pattern as the uart you mentioned. Currently this does not seem to be the case, so when a driver is defined using macro "DEVICE_DEFINE" then there is no check involved so the code runs regardless of state of device tree. Sure, it may not get compiled if Kconfig is not enabling it, but it should not be instantiated and run if it is enabled in Kconfig but device tree node is disabled. Unless DT macro is used to check for status OKAY, it will run anyway which breaks expectation about things actually being disabled when device tree node is disabled.

erwango replied:
> You're right, I think they should also be guarded after all. @erwango are you aware of any reason why it is not done?

@gmarull No reason. This should be fixed indeed.

> Do we enable I2C and SPI on board level and provide device tree nodes that are enabled by default? Do we provide a generic overlay for sample that enables only aliased devices? Do we even put CONFIG*_STM32* in samples because that would require sample to know which board it is running on? Is that wrong? So which approach is the right one? Is it to enable all board peripherals at board level instead of in sample overlays?

@mkschreder I'm not sure to follow. I've described the approach we're using and some of the arguments behind this approach. If you have alternative to propose, don't hesitate to raise an RFC and provides more details.

mkschreder replied:
> @mkschreder I'm not sure to follow. I've described the approach we're using and some of the arguments behind this approach. If you have alternative to propose, don't hesitate to raise an RFC and provides more details.

It makes a bit more sense to keep it this way after seeing how many tests fail if I change it and reading the board porting guide. I'll post RFC if I come up with more concrete changes to be made. Thanks for your clarification.

gmarull commented on drivers/memc/memc_stm32.c:
```diff
@@ -59,5 +59,7 @@ static const struct memc_stm32_config config = {
 	.pinctrl_len = ARRAY_SIZE(pinctrl),
 };
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, st_stm32_fmc), okay)
```
```suggestion
#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
```

gmarull commented on drivers/memc/memc_stm32_sdram.c:
```diff
@@ -129,5 +129,7 @@ static const struct memc_stm32_sdram_config config = {
 	.banks_len = ARRAY_SIZE(bank_config),
 };
 
+#if DT_NODE_HAS_STATUS(DT_INST(0, st_stm32_fmc_sdram), okay)
```
In this case it would not be needed because Kconfig option selecting SDRAM already takes into account DT status, but it can still be added.

```suggestion
#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
```

gmarull commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1_defconfig:
```diff
@@ -23,6 +23,12 @@ CONFIG_PINMUX=y
+CONFIG_MEMC=y
+
+# enable SDRAM
+CONFIG_MEMC_STM32_SDRAM=y
```
As discussed, board should not enable any of the `MEMC` options. Test/sample takes care of that.

gmarull reviewed (changes requested):
Requested some changes, then I think we'll be good to go. Another thing that could be changed is the `memc` commits title, I think `drivers: memc:` prefix is better.

erwango commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -103,6 +103,9 @@ static inline int stm32_clock_control_on(const struct device *dev,
 		LL_AHB2_GRP1_EnableClock(pclken->enr);
 		break;
+	case STM32_CLOCK_BUS_AHB3:
+		LL_AHB3_GRP1_EnableClock(pclken->enr);
```
Thanks for this addition.
Would you mind adding the equivalent in `stm32_clock_control_off` and `stm32_clock_control_get_subsys_rate`?

erwango commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -24,4 +25,21 @@
 		};
 	};
+
+	fmc: memory-controller@a0000000 {
```
Nodes definitions should be added in "soc dtsi files" (stm32f405.dtsi and stm32f446.dtsi in current case).
Current file is a "package dtsi file" and it aims at providing the possible memory variants of a SoC. Evrything that is common across variants should be defined at the lower level.

erwango commented on boards/arm/stm32f429i_disc1/stm32f429i_disc1_defconfig:
```diff
@@ -23,6 +23,12 @@ CONFIG_PINMUX=y
 CONFIG_GPIO=y
 
+# enable FMC
+CONFIG_MEMC=y
```
Please follow [board porting recommendations](https://docs.zephyrproject.org/latest/guides/porting/board_porting.html?highlight=board%20porting#general-recommendations) and let this to application responsibility.

erwango reviewed (changes requested):
Thanks for this addition!
Some comments on top of existing

mkschreder commented:
> Requested some changes, then I think we'll be good to go. Another thing that could be changed is the `memc` commits title, I think `drivers: memc:` prefix is better.

Good point. Fixed.

erwango commented on dts/arm/st/f4/stm32f429Xi.dtsi:
```diff
@@ -6,6 +6,7 @@
 
 #include <mem.h>
 #include <st/f4/stm32f429.dtsi>
+#include <dt-bindings/memory-controller/stm32-fmc-sdram.h>
```
Please move this one to dts/arm/st/f4/stm32f405.dtsi as well (/stm32f4.dtsi) alternatively.

gmarull replied:
I guess it should be moved to every file defining fmc/sdram node.

mkschreder replied:
Actually a few more are supported: https://www.st.com/en/microcontrollers-microprocessors/stm32f4-series.html

erwango reviewed (changes requested):
One last change and we're good to go.

gmarull commented on dts/arm/st/f4/stm32f405.dtsi:
```diff
@@ -242,4 +242,20 @@
 		label = "OTGHS_FS_PHY";
 	};
+
+	fmc: memory-controller@a0000000 {
```
STM32F405 doesn't seem to have FMC according to CubeMX. Only F427/F437/F429/F439/F446/F469/F479 have the peripheral.

erwango replied:
@gmarull This is my bad. I looked at CMSIS definitions and FMC is defined there in F405/F407. Haven't crossed checked on CubeMx. Looking to Ref manual, FMC support is indeed limited to F42X/F43X on the "F405/07 Line".
@mkschreder Sorry for the trouble, would you mind moving this definition to F427/F437/F446/F469 SoC dtsi files ?

mkschreder committed:
drivers: clock_control: add AHB3 clock support

This is needed to be able to use FMC on stm32f4

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder commented:
@gmarull I think this one is ready now.

gmarull commented on dts/arm/st/f4/stm32f405.dtsi:
```diff
@@ -242,4 +242,20 @@
 		label = "OTGHS_FS_PHY";
 	};
+
+	fmc: memory-controller@a0000000 {
```
FMC is not available on `F405`. You can check supported SoC e.g.
```
grep -m1 -R fmc path/to/hal_stm32/dts/st/f4 | sort
```

gmarull commented on dts/arm/st/f4/stm32f405.dtsi:
```diff
@@ -242,20 +242,4 @@
 		label = "OTGHS_FS_PHY";
 	};
-
-	fmc: memory-controller@a0000000 {
```
This commit should not contain that

gmarull commented on dts/arm/st/f4/stm32f412.dtsi:
```diff
@@ -164,4 +165,20 @@
 		};
 	};
+
+	fmc: memory-controller@a0000000 {
```
This commit should not contain that, please use a separate commit.

gmarull replied:
F412/413 embed FSMC, != FMC. FSMC does not support SDRAM.

mkschreder committed:
dts: soc: stm32f412: add fmc support

Add fmc node to device tree.

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder committed:
dts: soc: stm32f413: add fmc support

Add fmc support to device tree

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder committed:
dts: soc: stm32f427: add fmc support

Add fmc support to device tree.

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder committed:
dts: soc: stm32f446: add fmc support

Add fmc support to device tree

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

mkschreder committed:
boards: stm32f429i_disc1: add fmc and SDRAM support

Add fmc and sdram support to board.

Signed-off-by: Martin Schröder <mkschreder.uk@gmail.com>

gmarull reviewed (changes requested):
Some SOC still need to be updated (e.g. 469, 437...).

mkschreder commented:
469 includes this file. About the others, shall we leave them out for now?

On Thu, 10 Dec 2020, 15:18 Gerard Marull-Paretas, <notifications@github.com>
wrote:

> *@gmarull* requested changes on this pull request.
>
> Some SOC still need to be updated (e.g. 469, 437...).
> ------------------------------
>
> In dts/arm/st/f4/stm32f412.dtsi
> <https://github.com/zephyrproject-rtos/zephyr/pull/30438#discussion_r540200278>
> :
>
> > @@ -164,4 +165,20 @@
>  			num-bidir-endpoints = <6>;
>  		};
>  	};
> +
> +	fmc: memory-controller@a0000000 {
>
> F412/413 embed FSMC, != FMC. FSMC does not support SDRAM.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/30438#pullrequestreview-549235048>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCGVCIR7CMZ2BZT7T2PXC3SUDKEZANCNFSM4UNOM4LQ>
> .
>

gmarull commented:
> 469 includes this file. About the others, shall we leave them out for now?

Some files may be shared by multiple SoC, but there's one specific for 469, where FMC stuff should go in this case.

gmarull commented:
@mkschreder have you been able to check last review comments?

FRASTM reviewed (approved):
some changes in the `drivers/clock_control/clock_stm32_ll_common.c` already exist
